lib/segment/src/spaces/metric.rs

```diff
@@ -15,5 +15,6 @@ pub trait Metric {
     ) -> ScoreType;
 
     /// Necessary vector transformations performed before adding it to the collection (like normalization)
-    fn preprocess(&self, vector: Vec<VectorElementType>) -> Vec<VectorElementType>;
+    /// Return None if metric does not required preprocessing
+    fn preprocess(&self, vector: &[VectorElementType]) -> Option<Vec<VectorElementType>>;
```

Comment:
This is a bit ugly interface now, but I think performance-wise it is more critical to avoid copying rather. The alternative could be to make metric as enum instead of trait or to expose ``require_proccessing``. The would require ``match`` over enum types on each invocation and the second would require extra ``if`` statement and runtime error if the ``if`` is missed.

Sadly returning ``&[VectorElementType])`` back is not possible, as the ``preprocess`` method could create a new object.